### ISSUE TITLE
Conv2d with Learnable Delays

### ISSUE THREAD

**Issue by RaghavSinghal-10** [Thvnvtos/SNN-delays#5]
Hi,

I have been trying to implement Conv2d layers with learnable delays.

This is my attempt, setting, and reasoning till now:

We have input of the form: [batch_size (BS), C_in, time_steps (T), H_in, W_in].
We will use a Dcls3d layer (since we need to give in a 5d batched input).

Now, we will pass on the arguments as follows:
```
dilated_kernel_size = (max_delay, x, y) where x, y are the kernel_sizes used in normal conv2d
padding = (delay, x_pad, y_pad) where x_pad, y_pad are the padding used in normal conv2d
in_channels = C_in
out_channels = C_out
kernel_count = x*y?
```

We thus obtain:
```
m.weight: [C_out, C_in, kernel_count]
m.P: [3, C_out, C_in, kernel_count]
```

However, shouldn't m.P and m.weight have the same shapes?

There is something wrong with the implementation.

Could you kindly point out what seems wrong, and mention how to implement it correctly?

Thanks!

**Comment by tmasquelier** [comment]
Hello,
Thanks for your interest in our work!
You can indeed use DCLS3d, and then the temporal dimension of the 3D kernel corresponds to the delays.
The kernel count Kc is the number of elements in the kernel. It is a free parameter. It doesn't have to be equal to the x*y.
Each element of the kernel has one weight (thus the number of weights is Kc * Cin * Cout) and one 3D position (x,y,t) (thus the number of positions is 3 * Kc * Cin * Cout, stored in the P tensor).
All clear now?
Do not hesitate if you have other questions.
Cheers,

Tim Masquelier

**Comment by RaghavSinghal-10** [comment]
Thanks so much for the response!

I still have some queries:

Say I want to fit in a Conv2d layer with learnable delays instead of a generic Conv2d layer. Is this the right strategy to follow:

Normal Conv2d: Conv2d(2, 12, 5)

With learnable delays: Dcls3d(in_channels=2, out_channels=12, kernel_count=25, stride=1, padding=0, 
                              dilated_kernel_size=(self.max_delay,5,5), version='gauss')
                              
I make the input of the form [bs, c, t, h, w]; I pad the time dimension with a proper number of zeros and then apply Dcls3d.

Is there a Kc number that is a "good" number?

Thanks a lot for the help!

**Comment by tmasquelier** [comment]
Yes, you can try what you wrote, it makes sense.

Unfortunately, you'll have to search for a good Kc manually...

Rereading your issue, I think what you would like ideally is to learn only the positions along the temporal dimension (i.e. delays), while using the regular dense grid x*y for the spatial positions?
In the early days of DCLS we had a version called DCLS3-1d that did exactly that!
@K-H-Ismail : do you think it could be resuscitated?

**Comment by RaghavSinghal-10** [comment]
Yes, that's right, I would like to a normal Conv2d along the spatial dimensions, using the dense x*y grid, and then learn positions along the temporal dimension (which would be the delay) corresponding to each weight. In this setting, the number of weights would be equal to the number of positions.

I tried implementing the strategy that I described earlier on a simple dataset (NMNIST), and it doesn't seem to work very well.

On an unrelated note, although I know y'all implemented linear layers with learnable delays in your paper, do you think it makes sense to have Conv2d layers to do the same (so they can be used in vision datasets as well)? 

It ideally should benefit vs regular Conv2d right, atleast if there is a decently strong temporal dependence?

**Comment by tmasquelier** [comment]
Yes, I think Conv2d + delays are worth exploring for visual tasks.
I think delays will only help with temporal tasks, and I don't think NMNIST fits in that category; I would suggest action recognition, for example.

So, your experiment with DCLS3d did not work?
Maybe it's because the default init for positions (a normal law centered on the middle of the kernel) is not a good one for delays. I would instead use a uniform distribution to spread them in the whole temporal dimension, like we did in our delay paper (for the spatial coordinates you can keep the normal law). Let us know if you need help to do that.

I've just talked to @K-H-Ismail , and he is going to resuscitate DCLS3-1d, but this will take some time!

**Comment by K-H-Ismail** [comment]
Your suggestion have been added following this [issue](https://github.com/K-H-Ismail/Dilated-Convolution-with-Learnable-Spacings-PyTorch/issues/4). For future issues concerning DCLS method, please send them to [DCLS/issues](https://github.com/K-H-Ismail/Dilated-Convolution-with-Learnable-Spacings-PyTorch/issues) so I can see them.

You can get the latest dcls-0.1.0 release by doing 
```
pip3 install --upgrade --force-reinstall dcls
```
Currently, only the Dcls3_1d method is available. The last dimension of the input is always where positions are learned. Typically, for your case the last dimension of the input should be time t, so make sure you permute your input accordingly. Please see the [Usage](https://github.com/K-H-Ismail/Dilated-Convolution-with-Learnable-Spacings-PyTorch#usage) section for a use case.

**Comment by tmasquelier** [comment]
@K-H-Ismail : many thanks, that was fast!

@RaghavSinghal-10 : please try it and let us know if it works! Based on your previous post, the parameters you want to use are:
n_channels=2
out_channels=12
dilated_kernel_size = max_delay
dense_kernel_size=(5, 5)
kernel_count=1

This leads to 2x12x5x5 trainable weights, each associated with one learnable delay and one learnable sigma.

Note: if the final goal is to target a neuromorphic chip you may want to use decreasing sigmas (non-learnable) during training as we did in our paper and then round to the nearest integer delay at the end.

**Comment by RaghavSinghal-10** [comment]
@tmasquelier @K-H-Ismail 

Thanks a lot! This, indeed, was super fast!

@tmasquelier Yes, I will use non-learnable decreasing sigmas and proper init for the positions.

@K-H-Ismail, I will add future issues to the right channel moving forward.

Once again, thanks a lot for your help! I will let ya'll know how it goes!
